src/util/sandbox/test_sandbox.c

```diff
@@ -0,0 +1,397 @@
+        FD_LOG_WARNING(( "expected fd_sandbox_change_user to exit gracefully but " "stopped by signal %s", strsignal(WSTOPSIG(wstatus)) ));
+        return -1;
+      }
+      return 0;
+
+  } else { // child
+      struct passwd * uinfo = getpwnam("nobody");
+      if ( FD_UNLIKELY ( !uinfo ) )
+        FD_LOG_ERR(( "getpwnam: %s", strerror( errno ) ));
+      
+      fd_sandbox_change_user(uinfo);
+
+      struct passwd *current_user = getpwuid(getuid());
+      if ( FD_UNLIKELY ( strcmp( "nobody", current_user->pw_name ) ) ) 
+        FD_LOG_ERR(( "expected current user to be 'nobody'" ));
+      
+      exit(0);
+  }
+
+  return -1;
+}
+
+/* test_should_set_resource_limits ensures that, after calling `fd_sandbox_set_resource_limits`,
+   the set limits cannot be exceeded. */
+int
+test_should_set_resource_limits() {
+  pid_t pid, w;
+
+  if (( pid = fork() )) { // parent
+    int wstatus;
+    w = waitpid(pid, &wstatus, WUNTRACED);
+    if ( FD_UNLIKELY ( w == -1 ) ) {
+      FD_LOG_WARNING(( "waitpid: %s", strerror( errno ) ));
+      return -1;
+    }
+    if ( FD_UNLIKELY ( WIFEXITED( wstatus ) && WEXITSTATUS( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_set_resource_limits to exit gracefully but " "exited with non-zero status, status=%d", WEXITSTATUS(wstatus) ));
+      return -1;
+    } else if ( FD_UNLIKELY ( WIFSIGNALED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_set_resource_limits to exit gracefully but " "killed by signal %s", strsignal(WTERMSIG(wstatus)) ));
+      return -1;
+    } else if ( FD_UNLIKELY ( WIFSTOPPED( wstatus ) ) ) {
+      FD_LOG_WARNING(( "expected fd_sandbox_set_resource_limits to exit gracefully but " "stopped by signal %s", strsignal(WSTOPSIG(wstatus)) ));
+      return -1;
+    }
+    return 0;
+
+  } else { // child
+    // open should not fail
+    int fd = open("/etc/passwd", O_RDONLY);
+    if ( FD_UNLIKELY ( fd == -1 ) ) {
+      FD_LOG_WARNING(( "open: %s", strerror(errno) ));
+      return -1;
+    }
+    close(fd);
+
+    fd_sandbox_set_resource_limits(3U);
+
+    // open should fail
+    fd = open("/etc/passwd", O_RDONLY);
+    if ( FD_UNLIKELY ( fd == -1 && errno != EMFILE ) ) {
+      FD_LOG_WARNING(( "open: expected to fail for -24: %s", strerror(errno) ));
+      return -1;
+    } else if ( FD_UNLIKELY ( fd != -1 ) ) {
+      FD_LOG_WARNING(( "open expected to fail for -24 'Too many open files' but succedded" ));
+      exit(1);
+    }
+    exit(0);
+  }
+
+  return -1;
+}
+
+/* test_setup_netns ensures that, after calling `fd_sandbox_setup_netns`,
+   the process' view of the network interfaces is limited to the expected. */
+int
+test_setup_netns() {
+    pid_t pid, w;
+    
+    // Pick a test ns name.
+    char nsname[ 20 ];
+    snprintf(nsname, 20, "fdtestsbx-%u", getpid() );
+    printf("newname: %s\n", nsname);
+
+    // Create the new ns.
+    char cmd[ 64 ];
+    snprintf( cmd, 64, "ip netns add %s", nsname );
+    system(cmd);
```

Comment:
![nasty](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExMDdjM2NmMDgwZmE2OTE4NmI1MTA0NThlNTE2ZWQyM2I3OWYyYTNmZSZlcD12MV9pbnRlcm5hbF9naWZzX2dpZklkJmN0PWc/3o6gEbkoTAsbQkd5As/giphy.gif)

I love it :)